### ISSUE TITLE
When a protocol is defined in a stub, classes implementing that protocol are not checked for correctness

### ISSUE THREAD

**Issue by gracepetryk** [microsoft/pyright#6946]
**Describe the bug**
When implementing protocols defined in stubs, the implementing class is not checked for correctness. I first noticed this bug when implementing the `Sequence` protocol from `collections.abc`, but the bug also appears for simpler protocols such as `Container`.

**Code or Screenshots**
```python
from .typings.stub import SupportsSomething as SupportsSomething_stub
from .something_protocol import SupportsSomething as SupportsSomething_py


# no error raised
class ImplementsSomethingFromStub(SupportsSomething_stub):
    pass


# error raised as expected
class ImplementsSomethingFromPy(SupportsSomething_py):
    pass
```

The stub and normal python file have the same contents:
```python
from abc import abstractmethod
from typing import Protocol


class SupportsSomething(Protocol):
    @abstractmethod
    def something(self) -> int: ...
```
<img width="716" alt="Screenshot 2024-01-09 at 1 32 37 PM" src="https://github.com/microsoft/pyright/assets/5047809/bd1d7eb8-dc9e-43fe-9691-540bb081fa36">

**VS Code extension or command-line**
Language server in neovim using pyright `1.1.345`. Bug also appears in vscode.

**Comment by erictraut** [comment]
Thanks for the bug report.

This gets into an area of the type specification that is under-specified. It's an area that I'm hoping we'll get some clarity on in the near future.

Within a stub file, it's not clear how a type checker should differentiate between a protocol method that has a default implementation and one that does not, since all methods in a stub have an implementation of `...`. It appears that the typeshed stubs use `@abstractmethod` to differentiate between these two cases, as you've done in your code sample. That's a reasonable approach even though it's not documented anywhere in the typing spec.

I've updated pyright accordingly.

**Comment by gracepetryk** [comment]
I checked out your fix and I think I found another bug, though this may be an issue with typeshed. For types such as `Sequence` that inherit from a protocol but do not have `Protocol` as one of their base classes, no error is raised when there are unimplemented abstract methods in a final annotated class. I'm not able to replicate this with a simple example stub. This excerpt from [PEP-544](https://peps.python.org/pep-0544/#merging-and-extending-protocols) makes me think that pyright should complain in this case, at least when the implementing class is annotated with `@final`:
> Subclassing a protocol class would not turn the subclass into a protocol unless it also has typing.Protocol as an explicit base class. Without this base, the class is “downgraded” to a regular ABC that cannot be used with structural subtyping.

stub:
```python
from abc import abstractmethod
from typing import Protocol


class SupportsSomethingStub(Protocol):
    @abstractmethod
    def something(self, arg) -> int: ...


class AddsBehaviorStub(SupportsSomethingStub):
    def return_int(self, arg: int) -> int:
        return arg
```
example:
```python
from collections.abc import Collection, Sequence
from typing import Container, final

from .typings.stub import AddsBehaviorStub

# raises error as expected
@final
class ImplementsSomethingWithAddOne(AddsBehaviorStub):
    pass

# no error raised.
@final
class ImplementsSequence(Sequence[int]):
    pass


# Raises error as expected.
class ImplementsContainer(Container[int]):
    pass
```

**Comment by erictraut** [comment]
I think this one is debatable. If a class explicitly derives from a protocol but does not derive from `Protocol` itself, then it is considered a regular class, and it is responsible for implementing all methods of the protocol.

If the class also derives from `ABC`, pyright then treats it as a regular abstract class. You will get the behavior you're looking for if you change `AddsBehaviorStub` (or similarly change `Sequence`) to:

```python
class AddsBehaviorStub(SupportsSomething, ABC): ...
```

It looks like mypy's behavior is in accordance with what you're proposing here, and that behavior seems reasonable to me. If you'd like to open a new feature request in pyright's issue tracker, I would be open to adding this functionality to pyright.

I still think that the spec is not very clear on this point. If you'd like to propose a clarification or amendment to the typing spec, here are [the steps you could take](https://github.com/python/typing-council#decisions).

**Comment by gracepetryk** [comment]
`AddsBehaviorStub` already has the behavior I'm looking for, it's only the `collections.abc` stubs that don't throw an error as far as I can tell. I'm going to submit an issue to typeshed to add `ABC` as a subclass for all non-protocol types, since that seems more accurate to me based on the runtime behavior of `collections.abc`:

```
>>> from collections.abc import Sequence
>>> Sequence.__class__
<class 'abc.ABCMeta'>
```

**Comment by gracepetryk** [comment]
https://github.com/python/typeshed/issues/11266

**Comment by erictraut** [comment]
This is included in pyright 1.1.346, which I just published. It will also be included in a future release of pylance.
